### PR TITLE
Update JSON Dictionary spec

### PR DESCRIPTION
**Pull request by thomas-bc**
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

- Renamed `Type Name` to `Type Descriptor`
- Introduced `Enumerated Constant Descriptor` and `Struct Member Descriptor` for sub-dictionaries that describe each concept
- Renamed `description` to `annotation` throughout in the spec and added it where it was missing
- Clarified some of the Description/Option tables to avoid repetition
- Moved `Parameters` section after Events to respect the _CTEP_ (Commands>Telemetry>Events>Parameter) order

## Rationale

Updating the spec with regards to feedback from inputting into the GDS / build system

## Testing/Review Recommendations

Will be **drastically** easier to review with the "Display Rich Diff" option in the changed files
image

### CODE REVIEW

File: docs/Design/fpp-json-dict.md [hunk 1]
```diff
@@ -178,118 +149,145 @@
...
+            "value" : 1
+        },
+        {
+            "name" : "MAYBE",
+            "value" : 2,
+            "annotation": "The cat would know"
+        },
+    ],
+    "default": "M.Status.MAYBE",
+    "annotation": "Schrödinger's status"
```
[the remark this answers is not in the document]
Comment (thomas-bc) [reply docs/Design/fpp-json-dict.md:262]: @jsoref it appears there's a bug in spelling with the `ö` character?

File: docs/Design/fpp-json-dict.md [hunk 2]
```diff
@@ -178,118 +149,145 @@
...
+| ----- | ----------- | ------- | -------- |
+| `name` | Name of the enumerated constant | **String** | true |
+| `value` | Value associated to the enumerated constant | **Number** | true |
+| `annotation` | User-defined annotation for the enumerated constant | **String** | false |
 
 Example FPP model with JSON representation:
 ```
-module M {
+module M1 {
+    @ Schrödinger's status
```
[the remark this answers is not in the document]
Comment (thomas-bc) [reply docs/Design/fpp-json-dict.md:228]: 😭

[the remark this answers is not in the document]
Comment (thomas-bc) [reply docs/Design/fpp-json-dict.md]: image

File: docs/Design/fpp-json-dict.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomas-bc) [reply docs/Design/fpp-json-dict.md]: Seems like a bug with `ö` - will add Schr to `expect.txt` to circumvent

[the remark this answers is not in the document]
Comment (thomas-bc) [reply docs/Design/fpp-json-dict.md]: Or could have just went with Schrodinger's...

**Comment by bocchino** [comment]
FPP annotations must be ASCII character strings. You could replace Schrödinger with Schroedinger.

File: docs/Design/fpp-json-dict.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsoref) [reply docs/Design/fpp-json-dict.md]: check-spelling is happy to let you ignore things that look like foreign words... But if you don't want to do that, you wouldn't take its advice :). (It seems from below that you don't...)

We appreciated the meme ❤️.

**Review by bocchino — COMMENTED** [review]
Looks good! Just one comment. In the type descriptor for a string type, I think the `size` field should have the description "Maximum string size in bytes" and the field should be required.

**Comment by thomas-bc** [comment]
Agreed, fixed!

**Comment by thomas-bc** [comment]
Note to self: there's an issue with the rendering of some tables in the spec on the website: https://nasa.github.io/fprime/Design/fpp-json-dict.html

This could be the solution: https://github.com/orgs/community/discussions/23945#discussioncomment-3242285

**Comment by thomas-bc** [comment]
nvm: very likely has to do with whitespaces https://www.reddit.com/r/github/comments/15gexlw/comment/juik4y3

Fixed

File: docs/Design/fpp-json-dict.md [hunk 3]
```diff
@@ -131,11 +131,12 @@ Example: Type Descriptor for booleans
 ```
 
 ## String Type Descriptors
+
 | Field | Description | Options | Required |
 | ----- | ----------- | ------- | -------- |
 | `name` | **String** representing the FPP type name |  `string` | true |
 | `kind` | **String** representing the kind of type | `string` | true | 
-| `size` | **Number** of bytes supported by the data type | **Number** in the range [0, 2<sup>31</sup>) | false |
+| `size` | Maximum string size in bytes | **Number** in the range [0, 2<sup>31</sup>) | true |
```
Comment (bocchino) [inline docs/Design/fpp-json-dict.md]: Looks good. Maybe the size description should say `**Number** representing the maximum string size in bytes`, for consistency. Also, zero is an invalid string size. The min should be 1.

Comment (bocchino) [reply docs/Design/fpp-json-dict.md]: Actually the FPP spec does say zero is allowed, but this is wrong: https://fprime-community.github.io/fpp/fpp-spec.html#Types_String-Types. The fpp-check tool excludes zero, and this is right.
```
[I] bocchino@MT-306179 ~> fpp-check
struct S { s: string size 0 }
fpp-check
stdin:1.27
struct S { s: string size 0 }
                          ^
error: invalid string size 0
```

Comment (thomas-bc) [reply docs/Design/fpp-json-dict.md]: Fixed description and range starting at 1

Comment (bocchino) [reply docs/Design/fpp-json-dict.md]: I'll fix the error in the spec in my next PR to FPP.

**Review by bocchino — APPROVED** [review]
Looks good!

**Comment by thomas-bc** [comment]
This state of the JSON dictionary spec is in sync with fpp-to-dict=2.1.0a7 (https://pypi.org/project/fprime-fpp-to-dict/2.1.0a7/)

**Pull request merged**